common/async: add primitives for structured concurrency with c++20 coroutines

cbodley opened the pull request:
squashed/cleaned up commits that were initially part of the wip-coro-after-reef feature branch

more recently, a similar set of primitives were merged for stackful coroutines in https://github.com/ceph/ceph/pull/57188

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

cbodley commented:
@adamemerson i know you're carrying the original commits for this stuff in https://github.com/ceph/ceph/pull/56499, but i'd like to start merging prerequisites for metadata sync prototyping to unblock progress there

cc @smanjara

adamemerson commented on src/common/async/co_throttle.h:
```diff
@@ -0,0 +1,113 @@
+/// }
+/// \endcode
+template <boost::asio::execution::executor Executor>
+class co_throttle {
```
I'm not sure how much I like none of these being thread safe. I understand we don't want the overhead when we don't need to pay for it, but the lack of thread safety makes some things a bit difficult. I couldn't figure out how to get `co_throttle` to work with datalog recovery in a way that didn't crash on cancellation, for example/

Could we maybe have a template parameter to determine it?

cbodley replied:
i'm not sure where the gap is with cancellation. i added a unit test for both co_throttle and co_spawn_group that trigger cancellation from a different thread, and neither thread-sanitizer nor valgrind complain about data races. can you suggest any changes to that test to make it more like the datalog case?

cbodley replied:
when designing this stuff, i admittedly had a very specific use case in mind: allowing a parent coroutine to spawn and manage several child coroutines in a way that guarantees they can safely access memory on the parent's stack

as long as the co_throttle/spawn_group is owned by the parent stack, cancellation of the parent will cause the stack to unwind and the co_foo to be destroyed, which cancels any child coroutines before they have another chance to access the parent stack

here it's crucial that the parent coroutine is running on the same executor as the children so that this child cancellation is synchronous. if they're on different executors, the child cancellations will get dispatched asynchronously, leaving a window where they can race to access the parent stack that's gone away

i can certainly understand the desire to generalize from 'parent coroutine' to 'parent class', but it's very hard to guarantee that a class destructor will run on a specific executor. in the case of Datalog, rgw's shutdown is likely running outside of our io_context_pool entirely

in terms of thread safety, it should be possible for the class destructor to safely trigger cancellation of the children. but we can't guarantee that the child coroutines can safely access member variables or call member functions/coroutines which is a major limitation

cbodley replied:
the `co_throttle` interface does a better job of enforcing this guarantee, because `co_throttle::spawn()` wraps the call to `asio::co_spawn()` so always passes the correct executor

`co_spawn_group` should do the same, instead of functioning as a `CompletionToken` for manual calls to `asio::co_spawn()`. i'd thought that was a more flexible design, but not at the cost of safety

adamemerson commented on src/common/async/parallel_for_each.h:
```diff
@@ -0,0 +1,86 @@
+              std::sentinel_for<Sentinel, Iterator> &&
+              std::same_as<AwaitableT, void> &&
+              boost::asio::execution::executor<AwaitableExecutor>)
+auto parallel_for_each(Iterator begin, Sentinel end,
```
This range stuff, at least when I was trying to use it, makes the Clang14 used in make check fail. Though my preference would be to get rid of clang14 and just switch back to GCC if we can't get a newer clang.

cbodley commented:
jenkins test make check

cbodley commented:
jenkins test windows

cbodley commented:
@adamemerson could you please review the new squash commits?

* changed the `co_spawn_group` interface to make it impossible for child coroutines to run on a different executor
* added test cases for `co_spawn_group` and `co_throttle` that trigger cancellation from an arbitrary thread (verified with tsan)

cbodley commented:
> changed the `co_spawn_group` interface to make it impossible for child coroutines to run on a different executor

the `spawn_throttle` class that already merged in https://github.com/ceph/ceph/pull/57188 had the same issue, so i made the corresponding changes to that in https://github.com/ceph/ceph/pull/58798

adamemerson commented:
jenkins test make check arm

adamemerson commented:
jenkins test make check arm64

cbodley commented:
https://jenkins.ceph.com/job/ceph-windows-pull-requests/44181/

> [2024-07-24T20:00:00.000Z] [googletest] ceph_test_librbd failed. Error: Command returned non-zero code(1): "cmd /c 'C:\ceph\ceph_test_librbd.exe --gtest_output=xml:C:\workspace\test_results\out\ceph_test_librbd\ceph_test_librbd_results.xml  >> C:\workspace\test_results\out\ceph_test_librbd\ceph_test_librbd_results.log 2>&1'".
...
ceph_test_librbd.TestLibRBD.WriteZeroesThickProvision
